lib/rrule/rule.rb

```diff
@@ -94,39 +78,43 @@ def all_until(end_date: nil, limit: nil)
 
         results_with_time = generator.combine_dates_and_times(possible_days_of_year, timeset)
         results_with_time.sort.each do |this_result|
-          if end_date
-            if this_result > end_date
-              return result
-            end
-          end
-
-          if options[:until]
-            if this_result > options[:until]
-              return result
-            end
-            result.push(this_result)
-          elsif this_result >= dtstart
-            total += 1
-            if options[:count]
-              count -= 1
-              result.push(this_result)
-              return result if count == 0
-            else
-              result.push(this_result)
-            end
-          end
-
-          return result if limit && result.size == limit
+          next if this_result < dtstart
+          return if options[:until] && this_result > options[:until]
```

Comment:
@nerdrew You're right, `StopIteration` isn't necessary here and `return` is probably better, since it doesn't require anybody reading the code to know about the special behavior of raising `StopIteration` within a `loop`.